ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/StateMachine.hs

```diff
@@ -0,0 +1,529 @@
+
+  perform _ (Init secParam) _ = do
+    Environment _ _ chainDb mkArgs _ <- get
+    (ldb, od, clean) <- lift $ do
+      args <- mkArgs secParam
+      (ldb, od) <- openLedgerDB (argFlavorArgs args) chainDb (extLedgerDbConfig secParam) (argHasFS args)
+      pure (ldb, od, cleanup args)
+    put (Environment ldb od chainDb mkArgs (closeBackingStore od >> clean))
+  perform _ (DoTruncate s) lk = do
+    Environment _ od _ _ _ <- get
+    lift $ truncateSnapshot od (lk s)
+  perform _ (DoDelete s) lk = do
+    Environment _ od _ _ _ <- get
+    lift $ destroySnapshot od (lk s)
+  perform _ GetState _ = do
+    Environment ldb _ _ _ _ <- get
+    lift $ atomically $ (,) <$> getImmutableTip ldb <*> getVolatileTip ldb
+  perform _ ForceTakeSnapshot _ = do
+    Environment ldb od _ _ _ <- get
+    lift $ do
+      void $ flushNOW od >> takeSnapshotNOW od Nothing
+      imm <- atomically $ getImmutableTip ldb
+      let pt = pointToWithOriginRealPoint $ castPoint @(ExtLedgerState TestBlock) @TestBlock $ getTip imm
+      pure
+         $ withOrigin Nothing (\p -> Just (DiskSnapshot (unSlotNo $ realPointSlot p) Nothing))
+         $ pt
+  perform _ (ValidateAndCommit n blks) _ = do
+      Environment ldb _ chainDb _ _ <- get
+      lift $ do
+        atomically $ modifyTVar (dbBlocks chainDb) $
+           repeatedly (uncurry Map.insert) (map (\b -> (blockRealPoint b, b)) blks)
+        withRegistry $ \rr -> do
+          vr <- validate ldb rr (const $ pure ()) BlockCache.empty n (map getHeader blks)
+          case vr of
+            ValidateSuccessful forker -> do
+              atomically $ modifyTVar (dbChain chainDb) (reverse (map blockRealPoint blks) ++)
+              atomically (forkerCommit forker)
+              forkerClose forker
+            ValidateExceededRollBack{} -> error "Unexpected Rollback"
+            ValidateLedgerError (AnnLedgerError forker _ _) -> forkerClose forker >> error "Unexpected ledger error"
+  perform state@(Model _ _ secParam) (DropAndRestore n) lk = do
+    Environment _ od chainDb _ clean <- get
+    lift $ do
+      atomically $ modifyTVar (dbChain chainDb) (drop (fromIntegral n))
+      closeBackingStore od
+      clean
+    perform state (Init secParam) lk
+  perform UnInit _ _ = error "Uninitialized model created a command different than Init"
+
+
+  postcondition (Model chain _ secParam, _) GetState _ (imm, vol) =
+    let volSt = either forgetLedgerTables (forgetLedgerTables . snd) (AS.head chain)
+        immSt = either forgetLedgerTables (forgetLedgerTables . snd) (AS.head (AS.dropNewest (fromIntegral $ maxRollbacks secParam) chain))
+    in do
+       counterexamplePost $ unlines [ "VolSt: ", show volSt
+                                    , "VolSut: ", show vol
+                                    , "ImmSt: ", show immSt
+                                    , "ImmSut: ", show imm
+                                    ]
+       pure $ volSt == vol && immSt == imm
```

Comment:
Why not check all states? It's also probably useful to have a stricter postcondition in general. The stricter the postcondition, the more interesting cases you might catch

Reply:
There is no way this could be true and other intermediate states not match. It is implicit by the fact that blocks validate on top of each other.

Reply:
Also this tests the observable part of the LedgerDB. To be precise, we could be OK just by checking the tip.

Reply:
For posterity, we should document a statement like this in the code. It's an important conclusion that isn't obvious to a reader. One option is to use GHC-style notes, i.e., something like:

```haskell
{- Note [Sufficient post-condition]
~~~~~~~~~~~~~~~~~~
Something something interesting interesting
-}
```